pipeline/trusted-resources/pkg/trustedtask/trustedtaskrun_test.go

```diff
@@ -0,0 +1,626 @@
+		Name:        "tr",
+		Namespace:   nameSpace,
+		Annotations: map[string]string{},
+	}
+
+	// service account
+	sa = &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: nameSpace,
+			Name:      serviceAccount,
+		},
+	}
+
+	k8sclient    = fakek8s.NewSimpleClientset(sa)
+	tektonClient = faketekton.NewSimpleClientset(ts, tsTampered)
+)
+
+func init() {
+	os.Setenv("SYSTEM_NAMESPACE", nameSpace)
+	os.Setenv("WEBHOOK_SERVICEACCOUNT_NAME", serviceAccount)
+}
+
+func TestVerifyResources_TaskSpec(t *testing.T) {
+	ctx := context.Background()
+
+	// Get singer
+	signer, err := getSingerFromFile(t, ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tr := v1beta1.TaskRun{
+		TypeMeta:   trTypeMeta,
+		ObjectMeta: trObjectMeta,
+		Spec: v1beta1.TaskRunSpec{
+			TaskSpec: &ts.Spec,
+		},
+	}
+
+	apiTaskRun := TrustedTaskRun{TaskRun: tr}
+	taskRunPass := apiTaskRun.DeepCopy()
+	taskRunContentTampered := apiTaskRun.DeepCopy()
+	taskRunUnsigned := apiTaskRun.DeepCopy()
+
+	// Sign task
+	trList := []*TrustedTaskRun{
+		taskRunPass,
+		taskRunContentTampered,
+	}
+
+	for _, tr := range trList {
+		signatureEncoding := signTaskSpec(t, signer, tr.Spec.TaskSpec)
+		tr.Annotations["tekton.dev/signature"] = signatureEncoding
+	}
+
+	// Tamper taskSpec data after signed.
+	taskRunContentTampered.Spec.TaskSpec = &tsTampered.Spec
+
+	tcs := []struct {
+		name    string
+		taskRun *TrustedTaskRun
+		wantErr bool
+	}{{
+		name:    "API Task Pass Verification",
+		taskRun: taskRunPass,
+		wantErr: false,
+	}, {
+		name:    "API Task Fail Verification with tampered content",
+		taskRun: taskRunContentTampered,
+		wantErr: true,
+	}, {
+		name:    "API Task Fail Verification without signature",
+		taskRun: taskRunUnsigned,
+		wantErr: true,
+	},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			err := tc.taskRun.verifyTask(ctx, k8sclient, tektonClient)
+			if (err != nil) != tc.wantErr {
+				t.Fatalf("verifyResources get err %v, wantErr %t", err, tc.wantErr)
+			}
+		})
+	}
+
+}
+
+func TestVerifyResources_OCIBundle(t *testing.T) {
+	ctx := context.Background()
+
+	// Get singer
+	signer, err := getSingerFromFile(t, ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create registry server
+	s := httptest.NewServer(registry.New())
+	defer s.Close()
+	u, _ := url.Parse(s.URL)
+
+	// Push OCI bundle
+	dig, err := pushOCIImage(t, u, ts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = pushOCIImage(t, u, tsTampered)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// OCI taskruns
+	otr := v1beta1.TaskRun{
+		TypeMeta:   trTypeMeta,
+		ObjectMeta: trObjectMeta,
+		Spec: v1beta1.TaskRunSpec{
+			TaskRef: &v1beta1.TaskRef{
+				Name:   "ts",
+				Bundle: u.Host + "/task/" + ts.Name,
+			},
+		},
+	}
+
+	OCITaskRun := TrustedTaskRun{TaskRun: otr}
+	taskRunPassOCIBundle := OCITaskRun.DeepCopy()
+	taskRunPassOCIBundleContentTampered := OCITaskRun.DeepCopy()
+	taskRunPassOCIBundleUnsigned := OCITaskRun.DeepCopy()
```

Comment:
nit: These variable names are a bit long and confusing -i.e. taskRun**Pass**OCIBundleUnsigned actually expects an error.
Perhaps rename to something like...

```suggestion
	unsigned := TrustedTaskRun{TaskRun: v1beta1.TaskRun{...}}
	
	signed := unsigned.DeepCopy()
	signed.Annotations["tekton.dev/signature"] = signOCIBundle(t, signer, dig)
	
	tampered := signed.DeepCopy()
	tampered.Spec.TaskRef.Bundle = u.Host + "/task/" + tsTampered.Name
```

This also helps keep object modifications contiguous for less context switching!

Reply:
Thanks! This is much better